tool/teleport/common/sftp.go

```diff
@@ -52,275 +75,506 @@ func (c compositeCh) Close() error {
 	return trace.NewAggregate(c.r.Close(), c.w.Close())
 }
 
-func onSFTP() error {
-	chr, err := openFD(3, "chr")
-	if err != nil {
-		return err
+// sftpHandler provides handlers for a SFTP server.
+type sftpHandler struct {
+	logger *log.Entry
+	events chan<- *apievents.SFTP
+}
+
+func newSFTPHandler(logger *log.Entry, events chan<- *apievents.SFTP) *sftpHandler {
+	return &sftpHandler{
+		logger: logger,
+		events: events,
 	}
-	defer chr.Close()
-	chw, err := openFD(4, "chw")
-	if err != nil {
-		return err
+}
+
+// OpenFile handles 'open' requests when opening a file for reading
+// and writing is desired.
+func (s *sftpHandler) OpenFile(req *sftp.Request) (_ sftp.WriterAtReaderAt, retErr error) {
+	defer s.sendSFTPEvent(req, retErr)
+
+	if req.Filepath == "" {
+		return nil, os.ErrInvalid
 	}
-	defer chw.Close()
-	ch := compositeCh{chr, chw}
-	auditFile, err := openFD(5, "audit")
+
+	return s.openFile(req)
+}
+
+// Fileread handles 'open' requests when opening a file for reading
+// is desired.
+func (s *sftpHandler) Fileread(req *sftp.Request) (_ io.ReaderAt, retErr error) {
+	defer s.sendSFTPEvent(req, retErr)
+
+	if req.Filepath == "" {
+		return nil, os.ErrInvalid
+	}
+	if !req.Pflags().Read {
+		return nil, os.ErrInvalid
+	}
+
+	return s.openFile(req)
+}
+
+// Filewrite handles 'open' requests when opening a file for writing
+// is desired.
+func (s *sftpHandler) Filewrite(req *sftp.Request) (_ io.WriterAt, retErr error) {
+	defer s.sendSFTPEvent(req, retErr)
+
+	if req.Filepath == "" {
+		return nil, os.ErrInvalid
+	}
+	if !req.Pflags().Write {
+		return nil, os.ErrInvalid
+	}
+
+	return s.openFile(req)
+}
+
+func (s *sftpHandler) openFile(req *sftp.Request) (*os.File, error) {
+	var flags int
+	pflags := req.Pflags()
+	if pflags.Append {
+		flags |= os.O_APPEND
+	}
+	if pflags.Creat {
+		flags |= os.O_CREATE
+	}
+	if pflags.Excl {
+		flags |= os.O_EXCL
+	}
+	if pflags.Trunc {
+		flags |= os.O_TRUNC
+	}
+
+	if pflags.Read && pflags.Write {
+		flags |= os.O_RDWR
+	} else if pflags.Read {
+		flags |= os.O_RDONLY
+	} else if pflags.Write {
+		flags |= os.O_WRONLY
+	}
+
+	f, err := os.OpenFile(req.Filepath, flags, defaultFilePerms)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	defer auditFile.Close()
 
-	// Ensure the parent process will receive log messages from us
-	utils.InitLogger(utils.LoggingForDaemon, log.InfoLevel)
+	return f, nil
+}
 
-	sftpEvents := make(chan *apievents.SFTP, 1)
-	sftpSrv, err := sftp.NewServer(ch, sftp.WithRequestCallback(func(reqPacket sftp.RequestPacket) {
-		event, ok := handleSFTPEvent(reqPacket)
-		if !ok {
-			// We don't care about this type of SFTP request, move on
+// Filecmd handles file modification requests.
+func (s *sftpHandler) Filecmd(req *sftp.Request) (retErr error) {
```

Comment:
It's a shame we basically have to reimplement these command handlers ourselves. Are there no "default handlers" somewhere in the sftp package we could wrap to add auditing?